libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,642 @@
+{
+    /* This treewalk finds not yet ionized particles within the radius of the black hole, ionizes them and
+     * adds an instantaneous heating to them. */
+    TreeWalk tw[1] = {{0}};
+
+    tw->ev_label = "HELIUM";
+    /* This could select the black holes to be made quasars, but we do it below.*/
+    tw->haswork = NULL;
+    tw->tree = tree;
+
+    /* We set Hsml to a constant in ngbiter, so this
+     * searches a constant distance from the halo.*/
+    tw->visit = (TreeWalkVisitFunction) treewalk_visit_ngbiter;
+    tw->ngbiter_type_elsize = sizeof(TreeWalkNgbIterBase);
+    tw->ngbiter = ionize_ngbiter;
+
+    tw->fill = (TreeWalkFillQueryFunction) ionize_copy;
+    tw->reduce = NULL;
+    tw->postprocess = NULL;
+    tw->query_type_elsize = sizeof(TreeWalkQueryBase);
+    tw->result_type_elsize = sizeof(TreeWalkResultBase);
+
+    struct QSOPriv priv[1];
+    priv[0].spin = init_spinlocks(PartManager->NumPart);
+    priv[0].fof = fof;
+    /* Ionization counters*/
+    priv[0].N_ionized = ta_malloc("n_ionized", int64_t, omp_get_max_threads());
+    memset(priv[0].N_ionized, 0, sizeof(int64_t) * omp_get_max_threads());
+
+    tw->priv = priv;
+
+    /* This runs only on one BH*/
+    if(qso_ind > 0)
+        treewalk_run(tw, &qso_cand[qso_ind], 1);
+    else
+        treewalk_run(tw, NULL, 0);
+
+    int64_t N_ionized = 0;
+    int i;
+    for(i = 0; i < omp_get_max_threads(); i++)
+        N_ionized += priv[0].N_ionized[i];
+
+    ta_free(priv[0].N_ionized);
+
+    free_spinlocks(priv[0].spin);
+
+    return N_ionized;
+}
+
+/* Sequentially turns on quasars.
+ * Keeps adding new quasars until need_more_quasars() returns 0.
+ */
+static void
+turn_on_quasars(double redshift, FOFGroups * fof, ForceTree * tree)
+{
+    int ncand = 0;
+    int * qso_cand = NULL;
+    int64_t n_gas_tot=0, tot_n_ionized=0, ncand_tot=0;
+    sumup_large_ints(1, &SlotsManager->info[0].size, &n_gas_tot);
+    double atime = 1./(1 + redshift);
+    double desired_ion_frac = gsl_interp_eval(HeIII_intp, He_zz, XHeIII, atime, NULL);
+    /* If the desired ionization fraction is above a threshold (by default 0.95)
+     * ionize all particles*/
+    if(desired_ion_frac > QSOLightupParams.heIIIreion_finish_frac) {
+        int i, nionized=0;
+        int64_t nion_tot=0;
+        #pragma omp parallel for reduction(+: nionized)
+        for (i = 0; i < PartManager->NumPart; i++){
+            if (P[i].Type == 0)
+                nionized += ionize_single_particle(i);
+        }
+        sumup_large_ints(1, &nionized, &nion_tot);
+        message(0, "HeII: Helium ionization finished, flash-ionizing %ld particles (%g of total)\n", nion_tot, (double) nion_tot /(double) n_gas_tot);
+    }
+
+    double rhobar = All.CP.OmegaBaryon * (3 * HUBBLE * All.CP.HubbleParam * HUBBLE * All.CP.HubbleParam)/ (8 * M_PI * GRAVITY) / pow(1 + redshift,3);
+    double totbubblegasmass = 4 * M_PI / 3. * pow(QSOLightupParams.mean_bubble, 3) * rhobar;
```

Comment:
Aren't the bubbles in overdense regions? If so, then mean_bubble is probably not the bubble's proper 'radius'. mean_bubble_radius_param?

Reply:
Well - at z=3 with a 30 Mpc bubble all bubbles are basically mean density